beacon_node/beacon_chain/src/validator_monitor.rs

```diff
@@ -487,8 +500,66 @@ impl<T: EthSpec> ValidatorMonitor<T> {
                         u64_to_i64(validator.withdrawable_epoch),
                     );
                 }
+
+                // Add non finalized missed blocks for the monitored validators for the current slot of the current epoch
+                let current_slot = state.slot();
+                self.missed_blocks.contains(&(current_epoch, i as u64, current_slot))
+                    .then(|| {
+                        metrics::inc_counter_vec(
+                            &metrics::VALIDATOR_MONITOR_MISSED_NON_FINALIZED_BLOCKS_TOTAL,
+                            &[current_epoch.to_string().as_str(), current_slot.to_string().as_str(), id],
+                        );
+                    });
+            }
+        }
+    }
+
+    fn add_validators_missed_blocks(&mut self, current_epoch: Epoch, state: &BeaconState<T>) {
+        // Determine missed (non-finalized) blocks (can contain false positives)
+        let range_of_slot = (T::slots_per_epoch() as usize - MISSED_BLOCK_LAG_SLOTS) - 1;
+        for n in 0..range_of_slot {
+            let slot = state.slot() - Slot::new(n as u64);
+            let prev_slot = slot - 1;
+
+            // safeguard if we are on a new chain, the slot - 1 can be negative
+            if prev_slot < 0 {
+                break;
+            }
+
+            // condition for missed_block is defined such as block_root == block_root - n
+            // where the proposer who missed the block is the proposer of the block at block_root - n
+            if let (Ok(block_root), Ok(prev_block_root)) = (state.get_block_root(slot), state.get_block_root(prev_slot)) {
+                if block_root == prev_block_root {
+                    let epoch = slot.epoch(T::slots_per_epoch());
+                    if let Ok(shuffling_decision_block) = state.proposer_shuffling_decision_root(epoch, *block_root) {
+                        // TODO: AI(Joel) Ask the team if we can use the beacon proposer cache without the mutex
+                        // mutex is implemented in the caller (BeaconChain), not in the cache implementation
+                        if let Some(proposer) = self.beacon_proposer_cache.get_slot::<T>(shuffling_decision_block, slot) {
```

Comment:
Ask the team about the cache.

It seems that the cache is used with a Mutex in the beacon chain. I am not sure if:
- We also want to add a mutex here when reading the cache? (why? is the internal state of the cache changes when reading a value such as last value queries which would require a lock?)
- Do we want a shared mutex then?
Would love some pointers here!

Reply:
I think we probably want to share the cache with the beacon chain, which could be done by using an `Arc<Mutex<BeaconProposerCache>>`. That way the `BeaconChain` and `ValdiatorMonitor` will use the same cache

Seeing as the validator monitor is less important than other parts of beacon chain processing we might also want to avoid locking in this loop. We could probably do 1-2 calls to `BeaconProposerCache::get_epoch` instead, and then use those values in the loop.